internal/redactor/redactor.go

```diff
@@ -0,0 +1,425 @@
+	// faster than _filtering_ all the needles by first byte.
+	needlesByFirstByte [256][]string
+
+	// For synchronising writes. Each write can touch everything below.
+	mu sync.Mutex
+
+	// Redacted output written to this writer.
+	dst io.Writer
+
+	// Intermediate buffer to account for partially-written non-secrets.
+	// (i.e. we began redacting in case we're in the middle of a secret, but
+	// we might not be).
+	buf []byte
+
+	// Current redaction states - if we have begun redacting a potential secret
+	// there will be at least one of these.
+	// nextStates is the next set of states. To avoid creating millions of new
+	// slices, Write alternates between these two.
+	states, nextStates []state
+
+	// The ranges in buf we must redact on flush.
+	redact []subrange
+}
+
+// New returns a new Redactor.
+func New(dst io.Writer, subst string, needles []string) *Redactor {
+	r := &Redactor{
+		dst:   dst,
+		subst: []byte(subst),
+
+		// Preallocate a few things.
+		buf:        make([]byte, 0, 65536),
+		states:     make([]state, 0, len(needles)),
+		nextStates: make([]state, 0, len(needles)),
+		redact:     make([]subrange, 0, len(needles)),
+	}
+	r.Reset(needles)
+	return r
+}
+
+// Write redacts any secrets from the stream, and forwards the redacted stream
+// to the destination writer.
+func (r *Redactor) Write(b []byte) (int, error) {
+	if len(b) == 0 {
+		return 0, nil
+	}
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	// The high level:
+	// 1. Append b to the buffer.
+	// 2. Search through b to find instances of strings to redact. Store the
+	//    ranges of redactions in r.redact.
+	// 3. Merge overlapping redaction ranges.
+	// 4. Write as much of the buffer as we can without spilling incomplete
+	//    matches.
+	//
+	// Step 2 is complicated by the fact that each Write could contain a partial
+	// secret at the start or the end. So a buffer is needed to hold onto any
+	// incomplete matches (in case they _don't_ match), as well as some extra
+	// state (r.states) for tracking where we are in each incomplete match.
+	//
+	// Step 4 (mostly in flushTo) only looks complicated because it has to
+	// alternate between unredacted and redacted ranges, *and* handle the case
+	// where we've chosen to flush to inside a redacted range.
+
+	prevBufLen := len(r.buf)
+
+	// 1. Append b to the buffer.
+	r.buf = append(r.buf, b...)
+
+	// 2. Search through b to find instances of strings to redact. Store the
+	//    ranges of redactions in r.redact.
+	for n, c := range b {
+		bufidx := n + prevBufLen // where we are in the whole buffer
+
+		// In the middle of redacting?
+		for _, s := range r.states {
+			// Does the needle match on this byte?
+			if c != s.needle[s.matched] {
+				continue
+			}
+
+			// It matched!
+			s.matched++
+
+			// Have we fully matched this needle?
+			if s.matched < len(s.needle) {
+				// This state survives for another byte.
+				r.nextStates = append(r.nextStates, s)
+				continue
+			}
```

Comment:
if we have a partial match on a write - ie we have a needle `foobar`, and we're writing `fomenting` to the redactor - does this mean that the `state` for the partial match on `fo` will stick around until we've processed the entire string that we're writing? i don't think this is a flaw, just confirming if i'm reading the code right - it's implied(?) by the code, but not explicit.

Reply:
It should start matching on `f`, continue matching on `o`, but then on `m`, the first `if c != s.needle[s.matched] { continue }` should drop it.

Reply:
right, gotcha - the bumped partial match is explicitly added to `nextPartialMatches` if the next byte matches a couple lines down from that, and if it isn't explicitly added to `nextPartialMatches`, it'll get dropped.